### PR TITLE
Refactor Public Inputs and add them to the transcript

### LINKED ISSUE #133: Add Public Inputs to the transcript.
A requirement for using the Fiat-Shamir heuristic is that the computation of the challenges must must include all public values from the zk-proof statement. Failing to do so could lead a malicious prover to be able to forge proofs.
More details  in [this blog post](https://blog.trailofbits.com/2022/04/13/part-1-coordinated-disclosure-of-vulnerabilities-affecting-girault-bulletproofs-and-plonk/) by Trail of Bits.

### PR DESCRIPTION
**Pull request by davidnevadoc**
In this PR we have added the Public Inputs to the transcript as the first step of the `prove` and `verify` algorithms. They have been added as the evaluations of the PI polynomial over the used domain (`n` field elements).

Close #133

### CODE REVIEW

**Comment by davidnevadoc** [comment]
This method may not be best way of solving the issue so discussions on alternative solutions are welcome!

File: plonk-core/src/proof_system/proof.rs
Comment (CPerezz) [inline plonk-core/src/proof_system/proof.rs]: Are the `pi` recieved sparse?? 
If so, we should see if we can avoid adding the zeros and filter them out.

File: plonk-core/src/proof_system/prover.rs
Comment (CPerezz) [inline plonk-core/src/proof_system/prover.rs]: I don't think we should append the dense repr, we should do the sparse (in case it doesn't affect the security) as the zeros will consume an insane amount of time.

**Review by CPerezz — CHANGES_REQUESTED** [review]

File: plonk-core/src/proof_system/prover.rs
Comment (davidnevadoc) [reply plonk-core/src/proof_system/prover.rs]: We could do something like this too
``` rust
for (key, val) in self.cs.public_inputs_sparse_store.iter() {
    let static_label = Box::leak(format!("pi{}", key).into_boxed_str());
    transcript.append(static_label.as_bytes(), val)
}
```
If we ensure that `public_inputs_sparse_store` has no zero values I think there would be no malleability issues.

Comment (CPerezz) [reply plonk-core/src/proof_system/prover.rs]: Well, while this works, can't we use `String` or similar?? Box leaking works but it's something better to avoid if possible.

Comment (lopeetall) [reply plonk-core/src/proof_system/prover.rs]: Can we serialize the sparse representation and append to the transcript as bytes?

Comment (ghost) [reply plonk-core/src/proof_system/prover.rs]: 1. I think we should make a PR to `merlin` so that the label doesn't need `'static` lifetime (I'm sure there was some good original reason for this, but it shouldn't be mandatory, because if the data is consumed by the hasher, it doesn't need to live that long)

2. If transcript performance is an issue, then maybe also make a PR to `merlin` to allow for Blake2b or other hashes (it looks like it might use Keccak right now...probably not the best choice for performance). The transcript also has to support Poseidon or other hashes if recursion will ever be used (as part of the transcript is verified in-circuit)

3. I think it should be fine to append a sparse repr instead of dense, but it has to be done in the correct way. The most important thing is that there is only one way to represent each PI in the transcript, and no two PI share the same repr in the transcript. As @davidnevadoc points out this requires that the sparse repr does not have any zero values.

**Comment by ghost** [comment]
Halo 2 also includes the circuit description in the transcript, which I think is another good level of defense in depth. In general, my understanding of the Trail of Bits recommendation is that pretty much everything the verifier has access to should go in the transcript.

File: plonk-core/src/proof_system/prover.rs
Comment (lopeetall) [reply plonk-core/src/proof_system/prover.rs]: I'm AFK or else I would just check this now...instead I'll leave a comment so I don't forget to mention it: has anyone checked PlookUp branch to see if the lookup table vectors get included in the transcript?

**Comment by ghost** [comment]
Per Alistair's suggestion on ZK study club telegram, it's not a bad idea to add the first element of the KZG SRS to the transcript as well. This is a bit tricky as it's not generic over the poly commit scheme

File: plonk-core/src/proof_system/prover.rs
Comment (davidnevadoc) [reply plonk-core/src/proof_system/prover.rs]: I agree with these suggestions but since `merlin` has not merged a PR in more than a year maybe we should create our own fork.

Comment (bhgomes) [reply plonk-core/src/proof_system/prover.rs]: Something to be wary of is iteration order. If you aren't using an ordered hash map then there are no guarantees on how the pi are returned from the map. You should use either an ordered map or perform a sort by index (public_input_indices) before inserting into the transcript.

Comment (davidnevadoc) [reply plonk-core/src/proof_system/prover.rs]: Finally went for a `BTreeMap` struct inside `PI` to ensure the values are returned  in order when appended to the transcript.

File: plonk-core/src/proof_system/proof.rs
Comment (davidnevadoc) [reply plonk-core/src/proof_system/proof.rs]: They usually are, so now only non-zero values will be appended to the transcript.

File: plonk-core/src/proof_system/prover.rs
Comment (hdevalence) [reply plonk-core/src/proof_system/prover.rs]: > I think we should make a PR to merlin so that the label doesn't need 'static lifetime (I'm sure there was some good original reason for this, but it shouldn't be mandatory, because if the data is consumed by the hasher, it doesn't need to live that long)

Ah, no, actually, the `'static` restriction is there [on purpose](https://merlin.cool/use/protocol.html#defining-a-transcript-protocol) to prevent exactly this type of (mis)use! 

The labels are supposed to be fixed, rather than input-dependent, because the labels are used for message framing for the data that's fed into the transcript.  Merlin uses a tag-length-value format, so the transcript is guaranteed to be parseable (and hence produce a non-malleable transcript hash) if the labels are all fixed and none of them are prefixes of each other.

In this case, the correct way to feed variable data into the transcript is to use two messages, one for the application-level label, and one for the application-level data, each with their own `'static` label at the transcript level.

> I agree with these suggestions but since merlin has not merged a PR in more than a year maybe we should create our own fork.

The upstream moved to https://github.com/zkcrypto/merlin/ ; I'm happy to merge PRs there, but not this one, since the design is trying to steer you away from doing this :)

Comment (ghost) [reply plonk-core/src/proof_system/prover.rs]: Ah, this is understandable (I hadn't realized the use of `'static` was connected to this property). You have a good point that the variable-length PIs have to be handled better here - this kind of thing is a bit of a challenge (no pun) because of circuit configurability at runtime.

However I still have doubts whether using `'static` for this purpose is really the best way to accomplish this. For example, the `Box::leak` code above causes this guarantee to break, without any indication at all that something went wrong (other than leaking memory - but the memory guarantee is now tied to the correctness). To get rustc to enforce these guarantees at compile time, perhaps something like enforcing the labels to be from an `enum` might be more effective.

Comment (bhgomes) [reply plonk-core/src/proof_system/prover.rs]: Because the actual circuit structure (and the entire proof system actually) can be configured at runtime, the notion of "compile-time" labels is just not correct; it is defined at the wrong abstraction level. In the case of this repo, the keys are fixed at _circuit compile time_ but not at _Rust compile time_. The only way a safety feature would be useful here is if you could constrain the labels to be fixed during circuit compilation (which has nothing to do with Rust compile-time data).

Comment (hdevalence) [reply plonk-core/src/proof_system/prover.rs]: > For example, the Box::leak code above causes this guarantee to break, without any indication at all that something went wrong (other than leaking memory - but the memory guarantee is now tied to the correctness). 

Uh, sure, I guess, but the use of `Box::leak` *in a loop over user-supplied data* to intentionally leak memory in the middle of a crypto implementation seems like a pretty clear indicator that something is wrong with the implementation, no?

> Because the actual circuit structure (and the entire proof system actually) can be configured at runtime, the notion of "compile-time" labels is just not correct; it is defined at the wrong abstraction level.

No, it isn't. You can of course write an application that doesn't include all of its data in its transcript, or does so incorrectly (which is how we ended up here). There's no way for the transcript construction to solve that for you. What it can, and does, solve for you, is ensure that you only have to worry about the application-level data, and not the low-level details of ensuring that that data is unambiguously encoded in the transcript hash.

Comment (bhgomes) [reply plonk-core/src/proof_system/prover.rs]: > Uh, sure, I guess, but the use of Box::leak in a loop over user-supplied data to intentionally leak memory in the middle of a crypto implementation seems like a pretty clear indicator that something is wrong with the implementation, no?

The keys are not _user-supplied data_ in the sense of _prover-supplied_ or _verifier-supplied_, but they are _compiler-supplied_ where _compiler_ is the user defining the circuit who is allowed to do anything. I agree that `Box::leak` is bad though, we need to get rid of that regardless of the `'static` issue.

> No, it isn't. You can of course write an application that doesn't include all of its data in its transcript or does so incorrectly (which is how we ended up here). There's no way for the transcript construction to solve that for you. What it can, and does, solve for you, is ensure that you only have to worry about the application-level data, and not the low-level details of ensuring that that data is unambiguously encoded in the transcript hash.

It's the fact that the labels are `'static` that's the problem. The labels used when specifying the transcript protocol do not need to be known at _Rust compile time_ to be correct. Handling the "low-level details" in this case needs to be done by this repo; the unambiguous encoding _is defined at Rust runtime_.

And there are actually ways to have the transcript help the _circuit compile time_ enforcement of the transcript (independent of _Rust compile time_), they are just not a feature of the `merlin` project. For the long-term usefulness of this repo, we'll need to extend the functionality and create a more general tool for transcripts.

Comment (joebebel) [reply plonk-core/src/proof_system/prover.rs]: > Uh, sure, I guess, but the use of `Box::leak` _in a loop over user-supplied data_ to intentionally leak memory in the middle of a crypto implementation seems like a pretty clear indicator that something is wrong with the implementation, no?

I agree it's an implicit indication - and in this case it did work exactly the way you intended, because none of us would have accepted code containing `Box::leak` as a good idea - but it also isn't explicit, in the sense that using `Box::leak` is entirely memory-safe and type-safe, so it is entirely possible to violate the API requirements at compile-time without violating any of `rustc`'s compile-time rules.

At minimum, I didn't initially understand the purpose of using `'static` here, and didn't recognize the functional consequences of `Box::leak`. It's possible I'm just an idiot, but in the unlikely event I'm not, it would be nice if it's possible to enforce the API requirements using the type system instead of the borrow checker, which would offer a stronger guarantee, because it could be more explicit, and also not-workaround-able. However, I'll admit that it's not obvious to me how to do it.

> It's the fact that the labels are 'static that's the problem. The labels used when specifying the transcript protocol do not need to be known at Rust compile time to be correct. Handling the "low-level details" in this case needs to be done by this repo; the unambiguous encoding is defined at Rust runtime.

I think the `'static` discussion was a bit of a red herring (sorry!) and it's better to leave `'static` as it is (unless it's replaced with something better). The issue is we have multiple levels of abstraction here, with each layer having it's own requirements and guarantees.

Specifically, we have `merlin` offering certain guarantees about the transcript, as long as certain requirements are met (as described by @hdevalence above and the `merlin` docs). Plonk also needs to offer similar guarantees to the users of the proving system (circuit designers, etc) as long as its own requirements are met. In the former case, we can ensure the requirements are met at `rustc` compile-time, whereas the latter needs to ensure they are met at run-time. 

As Henry said above, there could be a fixed label set for Plonk at compile time (thus meeting the `'static` requirement) and any circuit-level configuration is handled as application-level data. For example, there is one fixed label "pi" for public inputs and then the data for "pi" is itself a run-time created transcript of some kind encoding the structure of the inputs. 

Now, there is a theoretical issue, that if the application-level itself is trying to offer the same guarantees and generality as `merlin` does, then duplicating this at two layers of abstraction is at best wasteful and at worst doubling the vulnerability space. In this general case I think @bhgomes is probably correct that everything must be enforced at run-time anyway. However, at least in this initial case (for example, variable length public inputs) this is probably not necessary.

**Comment by CPerezz** [comment]
I agree with @hdevalence since leaking memory on purpose is clearly an indicator that something is not correct.

I don't really see any other way anyways. The only thing that would mitigate the leaking stuff is to hash up all of the PI's and add that to the transcript. I agree that would make the verifier work more. But at the same time only forces us to `leak` a single Box which contains all of the PI's hashed instead of each one of them.  Maybe we could even reserve a 32-byte array and unsafely modify it when PI's are passed so that no leaking happens.

Otherways we will need to allocate one byte-array slot for each PI which makes it much more complex..
WDYT @davidnevadoc @joebebel @bhgomes ?

**Comment by ghost** [comment]
> I don't really see any other way anyways. The only thing that would mitigate the leaking stuff is to hash up all of the PI's and add that to the transcript. I agree that would make the verifier work more. But at the same time only forces us to `leak` a single Box which contains all of the PI's hashed instead of each one of them. Maybe we could even reserve a 32-byte array and unsafely modify it when PI's are passed so that no leaking happens.

We should to do something like @hdevalence suggested, which means that user configured parts of the proving system (including input length) are handled separately in some way before passing to `merlin`. This isn't really a "how to code it?" question; it's more a question of how we ensure that the protocol and transcript are well-formed, which `merlin` will handle for us to some extent.

For example, there is another problem with the code (besides the `Box::leak` distraction) - it violates the property that no label is a prefix of another label. While not necessarily being related to the runtime vs compile time issue, this is a good example of why this is rather difficult to get correct. Every time the "protocol" changes (which in `merlin`-terms includes the label set) there is potential risk, so `merlin` is designed (correctly) to nudge everyone away from "runtime" protocol changes, even ones that are technically safe, and to handle runtime variation above the `merlin` layer.

So instead the "static protocol" (from `merlin`'s perspective) needs to be general enough to handle runtime variation. We have to verify that the whole proving system can handle all user configured variations anyway, so there is not really any downside (for now) to making sure that the `merlin` transcript protocol is also general enough.

(did I miss anything @hdevalence?)

In the specific case of what to do with this PR, maybe we should hash the public inputs to 32 or 64 bytes and use a single PI label (no leaking) for the hash. (EDIT: or @lopeetall's suggestion of serializing the whole PI) The circuit description should also be added to the transcript earlier, so there is no malleability of the input order, etc.

**Comment by CPerezz** [comment]
@joebebel makes sense. 

Note that for the ordering, the hash will take care too. As we store the PI in a `BTreeSet` so therefore the order is preserved and therefore, ir prevents any order maleability.

**Comment by lopeetall** [comment]
The only issue I see with serializing the sparse form of the PI is that we would need to provide a spec for the serialization so that other implementations of the Verifier serialize PI in the same way.

Instead, why don't we just use the SRS to get a binding commitment the PI polynomial and add it to the transcript using the existing Merlin API wrapper? It may not be the most efficient solution as it requires doing another MSM, but we can still take advantage of the sparseness of the PI by skipping all the zeros. And since all we need is a binding commitment we won't need to use an FFT before MSM. 

This method would use our existing tools without modification and is consistent with the rest of the protocol.

**Comment by bhgomes** [comment]
We can just leave it as is but add the gate index to the transcript separately:

```rust
for (key, val) in self.cs.public_inputs_sparse_store.iter() {
    transcript.append("pi_gate_index", key);
    transcript.append("pi_value", val);
}
```

**Comment by bhgomes** [comment]
> For example, there is another problem with the code (besides the Box::leak distraction) - it violates the property that no label is a prefix of another label. 

Why is prefixing an issue? The documentation on `merlin` says that it's a requirement of the transcript being "parseable". What does this mean? We aren't parsing anything here.

**Comment by CPerezz** [comment]
> The only issue I see with serializing the sparse form of the PI is that we would need to provide a spec for the serialization so that other implementations of the Verifier serialize PI in the same way.

In the verifier (Usually a Smart Contract) will probably be much more expensive to do the MSM than the hash. That's why I suggested that @lopeetall.

For now, we could start with something easy that solves the bug. And then move on into a better design.

Something like this as purposed by @davidnevadoc in private.
```rust
// For each PI != 0
transcript.append(b"pi_pos", &F::from(*pos));
transcript.append(b"pi_val", &F::from(val));
```

This is not the most optimal. But works for now. And we can discuss the approach we want to go for in a follow-up issue.

**Comment by lopeetall** [comment]
Can we reuse those labels `pi_pos` and `pi_val` though?

From the Merlin website:

> The labels should be fixed, not defined at runtime, as runtime data should be in the message body, while the labels are part of the protocol definition. A sufficient condition for the transcript to be parseable is that the labels should be distinct and none should be a prefix of any other.

**Comment by ghost** [comment]
> Why is prefixing an issue? The documentation on `merlin` says that it's a requirement of the transcript being "parseable". What does this mean? We aren't parsing anything here.

The absolute requirement is that the transcript is binding to the messages in it, that is, it is not feasible to find a different transcript with different labels/messages that gives the same challenges. The property that `merlin` promises is that if labels are prefix free, then the transcript is unambiguously parseable (a mathematical definition - there is no computation to parse the transcript), and if the transcript is parseable then it is message binding. 

It could be that the transcript is message binding even if the labels are not prefix-free, but this requires an additional argument and can't be assumed from the guarantees that `merlin` provides. It's just much simpler to defer to `merlin`'s requirements.

> Instead, why don't we just use the SRS to get a binding commitment the PI polynomial and add it to the transcript using the existing Merlin API wrapper? It may not be the most efficient solution as it requires doing another MSM, but we can still take advantage of the sparseness of the PI by skipping all the zeros. And since all we need is a binding commitment we won't need to use an FFT before MSM.

Or, just serialize the PI polynomial? Though I think your original suggestion of serializing the PI directly is probably fine as well. I think the MSM is expensive.

**Comment by bhgomes** [comment]
> > Why is prefixing an issue? The documentation on `merlin` says that it's a requirement of the transcript being "parseable". What does this mean? We aren't parsing anything here.
> 
> The absolute requirement is that the transcript is binding to the messages in it, that is, it is not feasible to find a different transcript with different labels/messages that gives the same challenges. The property that `merlin` promises is that if labels are prefix free, then the transcript is unambiguously parseable (a mathematical definition - there is no computation to parse the transcript), and if the transcript is parseable then it is message binding.
> 
> It could be that the transcript is message binding even if the labels are not prefix-free, but this requires an additional argument and can't be assumed from the guarantees that `merlin` provides. It's just much simpler to defer to `merlin`'s requirements.

If this is the case then why are we letting the user specify any labeling scheme at all, just choose a prefix-free sequence of labels and use each one in turn. From an engineering perspective, I don't want to have to think about these safety requirements if there exists a protocol that takes care of it for me, especially when these labels don't actually help me accomplish anything inside of my application (in this case PLONK). This is the API I actually care about:

```rust
trait Transcript {
    fn append<M>(&mut self, message: M);
    fn challenge<C>(&mut self) -> C;
}
```

In which scenarios is knowing or specifying the labels manually an important feature to have?

**Comment by ghost** [comment]
> In which scenarios is knowing or specifying the labels manually an important feature to have?

Yeah, you are absolutely correct that from an implementation perspective, it is entirely possible to do without labels, or without `'static` lifetimes on the labels, etc etc. It's probably entirely possible to use the exact "pi{}" format labels too.

The reason it's important, though, is because "Frozen Heart" is just the latest in a long history of Fiat-Shamir implementation mistakes (mostly of the "omit data" kind). The idea is that we should write down - very explicitly - the exact Plonk interactive protocol being used, and then duplicate it exactly with `merlin`, down to the labels and everything. The goal is not only 0 mistakes in converting the interactive proof to non-interactive, but not even the possibility of mistakes.

**Comment by davidnevadoc** [comment]
After reading the discussion and learning about the correct usage of labels in `merlin` I believe that we should open a different PR to tackle this issue, as there are several parts of the code where they are misused. For example we are adding the commitment for the permutation with the label `z` and also adding the challenge used for lookup table RLC with the label `zeta`, which breaks the prefix rule. We are also using `Box::leak()`  in the same way I did in the first version of the PR when adding the evaluation of custom gate constraints.

Having the `merlin` requirements in mind I think the simplest solution is to serialize `PI` and add it to the transcript with the label "pi".  This struct is responsible of holding a one-to-one representation of each possible public input vector. In particular we use `BTreeMap` to guarantee the inputs are stored in order and we only store non-zero values, as zeros are implicit when no value is provided.

**Comment by hdevalence** [comment]
> After reading the discussion and learning about the correct usage of labels in `merlin` I believe that we should open a different PR to tackle this issue, as there are several parts of the code where they are misused.

The docs on [Transcript Protocols](https://merlin.cool/use/protocol.html) might be helpful; the recommended way to use the library is for each protocol to define a protocol-specific extension trait that maps the protocol-level transcript actions ("do a domain separation", "append a public key", "append a commitment to whatever", etc) operating on higher-level types into the underlying calls to the byte-oriented Merlin API.  So in this case, for instance, it might make sense to have a transcript protocol that handled the serialization and labeling, and presented higher-level methods to the rest of the crate.

File: plonk-core/src/circuit.rs
Comment (bhgomes) [inline plonk-core/src/circuit.rs]: stray comment?

File: plonk-core/src/constraint_system/composer.rs
Comment (bhgomes) [inline plonk-core/src/constraint_system/composer.rs]: Does this count the padding with the extra space needed for lookups?

File: plonk-core/src/constraint_system/helper.rs
Comment (bhgomes) [inline plonk-core/src/constraint_system/helper.rs]: Stray comment?

File: plonk-core/src/proof_system/pi.rs
Comment (bhgomes) [inline plonk-core/src/proof_system/pi.rs]: I would call this `PublicInput` or `PublicInputs`

File: plonk-core/src/proof_system/pi.rs
Comment (bhgomes) [inline plonk-core/src/proof_system/pi.rs]: Safer to use `derivative::Derivative` for `Clone`, `Debug`, `Hash`, `PartialEq`

File: plonk-core/src/proof_system/pi.rs
Comment (bhgomes) [inline plonk-core/src/proof_system/pi.rs:56]: In which case do you need this? This seems unsafe.

File: plonk-core/src/proof_system/pi.rs
Comment (bhgomes) [inline plonk-core/src/proof_system/pi.rs:71]: This does not prevent someone from doing the following:

```rust
self.insert(4, 0);
self.insert(4, 1);
```

or in the other order. Do we even want to allow this?

File: plonk-core/src/proof_system/pi.rs
Comment (bhgomes) [inline plonk-core/src/proof_system/pi.rs]: I think you need to accumulate as you go, this won't insert in the right place. Take for example the following structure

```
[v0, [v1, v2], v3, v4]
```

In this case the loops would put them at the following indices:

```
v0 -> 0 + 0 = 0  (element 0 of group 0)
v1 -> 1 + 0 = 1  (element 0 of group 1)
v2 -> 1 + 1 = 2  (element 1 of group 1)
v3 -> 2 + 0 = 2  (element 0 of group 2)
v4 -> 3 + 0 = 3  (element 0 of group 3)
```

In this case, we have a conflict at index `3`. Instead, you can just loop over the arrays and count as you go:

```rust
let mut count = 0;
for item in iter {
    for elem in item.to_field_elements().ok_or(Error::InvalidPublicInputValue)? {
        self.insert(init_pos + count, elem);
        count += 1;
    }
}
Ok(count)
```

File: plonk-core/src/proof_system/proof.rs
Comment (bhgomes) [inline plonk-core/src/proof_system/proof.rs]: not sure if "formally verify" is the right language here

**Review by bhgomes — CHANGES_REQUESTED** [review]
Just a few points, and a possible bug, otherwise LGTM.

File: plonk-core/src/circuit.rs
Comment (davidnevadoc) [reply plonk-core/src/circuit.rs]: yep 😅

File: plonk-core/src/proof_system/proof.rs
Comment (davidnevadoc) [reply plonk-core/src/proof_system/proof.rs]: I agree, in fact I would remove all the paragraph. WDYT?

Comment (bhgomes) [reply plonk-core/src/proof_system/proof.rs]: yeah let's get rid of it

File: plonk-core/src/proof_system/pi.rs
Comment (davidnevadoc) [reply plonk-core/src/proof_system/pi.rs:56]: This is related with the previous comment. It is used when the final size of the circuit is greater than the initially provided `expected_value`.  Once `gadget` has been called we know the real size of the circuit and we use this to ensure the size is updated if necessary.

File: plonk-core/src/proof_system/pi.rs
Comment (davidnevadoc) [reply plonk-core/src/proof_system/pi.rs:71]: You're right. I thought of this case but I didn't find a workaround for it as we don't keep track of `0`s. When we switch to the approach where we explicitly specify which positions are allowed to have a public input we can keep track of all insertions and prevent scenarios like this one.
The reverse would panic since the position check is done before the zero check.

File: plonk-core/src/constraint_system/composer.rs
Comment (davidnevadoc) [reply plonk-core/src/constraint_system/composer.rs]: This just depends on the `expected_size` provided at the initialization of the `StandardComposer`. If the user building the circuit knows that the lookup tables size will exceed the number of constraints he should use that value as `expected_size`. The value of the PI size will be updated with the correct size of the circuit once the circuit is finished (typically at the end of the `gadget` call).

File: plonk-core/src/proof_system/pi.rs
Comment (davidnevadoc) [reply plonk-core/src/proof_system/pi.rs]: Good point :+1:

**Review by CPerezz — APPROVED** [review]
LGTM. 

I like that refactor of the Public Inputs TBH. Although I think we can do more in the future with that and improve them much more.

As for the concerns of @bhgomes , they're legit. But at the same time, panicking due to a PI re-assignation is not something crazy. As it means you're not building a correct circuit. Which is not worth then compiling or making proofs for.

Would like the approval of @bhgomes and maybe one more before we merge TBH as this topic lead to a broad discussion and lots of members were involved.

**Review by bhgomes — APPROVED** [review]

**Review by LukePearson1 — APPROVED** [review]
LGTM. Feel free to merge

**Pull request merged**
